File: trunk/java/picocontainer/src/java/org/picocontainer/defaults/AbstractComponentAdapter.java
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the LICENSE.txt file.                                                     *
 *                                                                           *
 * Original code by                                                          *
 *****************************************************************************/
package org.picocontainer.defaults;

import org.picocontainer.ComponentAdapter;
import org.picocontainer.PicoContainer;
import org.picocontainer.PicoVisitor;

import java.io.Serializable;

/**
 * Base class for a ComponentAdapter with general functionality.
 * This implementation provides basic checks for a healthy implementation of a ComponentAdapter.
 * It does not allow to use <code>null</code> for the component key or the implementation, 
 * ensures that the implementation is a concrete class and that the key is assignable from the 
 * implementation if the key represents a type.   
 *  
 * @author Paul Hammant
 * @author Aslak Helles&oslash;y
 * @author Jon Tirs&eacute;n
 * @version $Revision$
 * @since 1.0
 */
public abstract class AbstractComponentAdapter implements ComponentAdapter, Serializable {
    private Object componentKey;
    private Class componentImplementation;
    private PicoContainer container;

    /**
     * Constructs a new ComponentAdapter for the given key and implementation. 
     * @param componentKey the search key for this implementation
     * @param componentImplementation the concrete implementation
     * @throws AssignabilityRegistrationException if the key is a type and the implementation cannot be assigned to.
     */
    protected AbstractComponentAdapter(Object componentKey, Class componentImplementation) throws AssignabilityRegistrationException {
        if (componentImplementation == null) {
            throw new NullPointerException("componentImplementation");
        }
        this.componentKey = componentKey;
        this.componentImplementation = componentImplementation;
        checkTypeCompatibility();
    }

    /**
     * {@inheritDoc}
     * @see org.picocontainer.ComponentAdapter#getComponentKey()
     */
    public Object getComponentKey() {
        if (componentKey == null) {
            throw new NullPointerException("componentKey");
        }
        return componentKey;
    }

    /**
     * {@inheritDoc}
     * @see org.picocontainer.ComponentAdapter#getComponentImplementation()
     */
    public Class getComponentImplementation() {
        return componentImplementation;
    }

    private void checkTypeCompatibility() throws AssignabilityRegistrationException {
        if (componentKey instanceof Class) {
            Class componentType = (Class) componentKey;
            if (!componentType.isAssignableFrom(componentImplementation)) {
                throw new AssignabilityRegistrationException(componentType, componentImplementation);
            }
        }
    }


    /**
     * @return Returns the ComponentAdapter's class name and the component's key.
     * @see java.lang.Object#toString()
     */
    public String toString() {
        return getClass().getName() + "[" + getComponentKey() + "]";
    }

    /**
     * @see org.picocontainer.ComponentAdapter#getContainer()
     */
    public PicoContainer getContainer() {
        return container;
    }

    /**
     * @see org.picocontainer.ComponentAdapter#setContainer(org.picocontainer.PicoContainer)
     */
    public void setContainer(PicoContainer picoContainer) {
        // Cater for alternate implementations of MutablePicoContainer.
        // ... that may leverage DefaultPicoContainer (wrap)
        if (this.container == null) {
            this.container = picoContainer;
        }
    }

    public void accept(PicoVisitor visitor) {
        visitor.visitComponentAdapter(this);
    }
}

File: trunk/dotnet/picocontainer/src/Defaults/AbstractComponentAdapter.cs
/*****************************************************************************
 * Copyright (C) PicoContainer Organization. All rights reserved.            *
 * ------------------------------------------------------------------------- *
 * The software in this package is published under the terms of the BSD      *
 * style license a copy of which has been included with this distribution in *
 * the license.txt file.                                                     *
 *                                                                           *
 * Idea by Rachel Davies, Original code by Aslak Hellesoy and Paul Hammant   *
 * C# port by Maarten Grootendorst                                           *
 *****************************************************************************/

using System;

using PicoContainer;

namespace PicoContainer.Defaults {

  public abstract class AbstractComponentAdapter : IComponentAdapter {
    private readonly object componentKey;
    private readonly Type componentImplementation;
    private IPicoContainer container;

    protected AbstractComponentAdapter(object componentKey, Type componentImplementation) {
      if(componentImplementation == null) {
        throw new NullReferenceException("componentImplementation");
      }
      this.componentKey = componentKey;
      this.componentImplementation = componentImplementation;
      CheckTypeCompatibility();
      CheckConcrete();
    }

    public object ComponentKey {
      get {
        if(componentKey == null) {
          throw new NullReferenceException("componentKey");
        }
        return componentKey;
      }
    }

    public Type ComponentImplementation {
      get {
        return componentImplementation;
      }
    }

    private void CheckTypeCompatibility() {
      Type componentType = componentKey as Type;
      if (componentType != null) {
        if (!componentType.IsAssignableFrom(componentImplementation)) {
          throw new AssignabilityRegistrationException(componentType, componentImplementation);
        }
      }
    }

    private void CheckConcrete() {
      if (componentImplementation.IsInterface || componentImplementation.IsAbstract) {
        throw new NotConcreteRegistrationException(componentImplementation);
      }
    }

    public override string ToString() {
      return this.GetType().Name+"["+ComponentKey+"]";
    }
    
    public IPicoContainer Container {
      get {
        return container;
      }

      set {
        this.container = value;
      }
    }

    public abstract object ComponentInstance{
      get;
    }

    public abstract void Verify();
  }
}
